crates/accelerate/src/synthesis/cnot_synthesis.rs

```diff
@@ -0,0 +1,188 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use hashbrown::HashMap;
+use ndarray::{s, Array1, Array2, ArrayViewMut2, Axis};
+use numpy::PyReadonlyArray2;
+use smallvec::smallvec;
+use std::cmp;
+
+use qiskit_circuit::circuit_data::CircuitData;
+use qiskit_circuit::operations::{Param, StandardGate};
+use qiskit_circuit::Qubit;
+
+use pyo3::prelude::*;
+
+/// Mutate a matrix inplace by adding the value of the ``ctrl`` row to the
+/// ``target`` row. If ``add_cols`` is true, add columns instead of rows.
+fn _add_row_or_col(mut mat: ArrayViewMut2<bool>, add_cols: &bool, ctrl: usize, trgt: usize) {
+    // get the two rows (or columns)
+    let info = if *add_cols {
+        (s![.., ctrl], s![.., trgt])
+    } else {
+        (s![ctrl, ..], s![trgt, ..])
+    };
+    let (row0, mut row1) = mat.multi_slice_mut(info);
+
+    // add them inplace
+    row1.zip_mut_with(&row0, |x, &y| *x ^= y);
+}
+
+/// This helper function allows transposed access to a matrix.
+fn _index(transpose: &bool, i: &usize, j: &usize) -> (usize, usize) {
```

Comment:
if this is a helper function for linear matrices, I would suggest to add it to synthesis/linear/utils.rs

Reply:
I don't think this is used anywhere else, so it might be cleaner to keep it private in this file -- if we find that we need it elsewhere I agree we should move it 🙂